packages/workflow/src/Workflow.ts

```diff
@@ -1421,6 +1424,74 @@ export class Workflow {
 			};
 		}
 	}
+
+	/**
+	 * Select the node to use as starter for a subworkflow execution.
+	 */
+	static selectSubworkflowStarter(nodes: INode[]) {
+		const found = nodes.find((n) => SUBWORKFLOW_STARTER_NODES.includes(n.type));
+
+		if (found) return found;
+
+		const starterNodesToDisplay = SUBWORKFLOW_STARTER_NODES.join(', ');
+
+		throw new ApplicationError(
+			`Missing node to start execution. Please ensure the workflow you are calling contains at least one of the following nodes: ${starterNodesToDisplay}`,
+		);
+	}
+
+	/**
+	 * Select the pinned activator node to use as starter for a manual execution.
+	 *
+	 * In a full manual execution, select the pinned activator that was first added
+	 * to the workflow, prioritizing `n8n-nodes-base.webhook` over other activators.
+	 *
+	 * In a partial manual execution, if the executed node has parent nodes among the
+	 * pinned activators, select the pinned activator that was first added to the workflow,
+	 * prioritizing `n8n-nodes-base.webhook` over other activators. If the executed node
+	 * has no upstream nodes and is itself is a pinned activator, select it.
+	 */
+	selectPinnedActivatorStarter(startNodeNames: string[]) {
+		const allPinnedActivators = this.findAllPinnedActivators();
+
+		const [firstPinnedActivator] = allPinnedActivators;
+
+		// full manual execution
+
+		if (startNodeNames?.length === 0) return firstPinnedActivator ?? null;
+
+		// partial manual execution
+
+		const [firstStartNodeName] = startNodeNames;
```

Comment:
This preserves existing behavior - if the partial manual execution has 2+ start nodes, we search only the zeroth start node's parents for a pinned activator. If we had 2+ start nodes without a common ancestor and so if we end up finding multiple pinned activators, we'd still need to return one to comply with existing usage, so I assume that picking the zeroth starter node here is correct.